acestep/training/path_safety.py

```
@@ -0,0 +1,97 @@
+        normalised = os.path.normpath(user_path)
+    else:
+        normalised = os.path.normpath(os.path.join(root, user_path))
+
+    # ── CodeQL-recognised sanitiser barrier ──
+    # ``normpath(…).startswith(safe_prefix)`` is the pattern that
+    # CodeQL's ``py/path-injection`` query treats as a sanitiser.
+    if not normalised.startswith(root + os.sep) and normalised != root:
+        raise ValueError(
+            f"Path escapes safe root: {user_path!r} "
+            f"(resolved to {normalised!r}, root={root!r})"
+        )
```

Ruff (0.15.0)

[warning] Avoid specifying long messages outside the exception class

(TRY003)